// knex/lib/helpers.js
/* eslint no-console:0 */

'use strict';

exports.__esModule = true;
exports.skim = skim;
exports.normalizeArr = normalizeArr;
exports.debugLog = debugLog;
exports.error = error;
exports.deprecate = deprecate;
exports.warn = warn;
exports.exit = exit;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _lodash = require('lodash');

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

// Pick off the attributes from only the current layer of the object.

function skim(data) {
  return _lodash.map(data, function (obj) {
    return _lodash.pick(obj, _lodash.keys(obj));
  });
}

// Check if the first argument is an array, otherwise uses all arguments as an
// array.

function normalizeArr() {
  var args = new Array(arguments.length);
  for (var i = 0; i < args.length; i++) {
    args[i] = arguments[i];
  }
  if (Array.isArray(args[0])) {
    return args[0];
  }
  return args;
}

function debugLog(msg) {
  console.log(msg);
}

function error(msg) {
  console.log(_chalk2['default'].red('Knex:Error ' + msg));
}

// Used to signify deprecated functionality.

function deprecate(method, alternate) {
  warn(method + ' is deprecated, please use ' + alternate);
}

// Used to warn about incorrect use, without error'ing

function warn(msg) {
  console.log(_chalk2['default'].yellow('Knex:warning - ' + msg));
}

function exit(msg) {
  console.log(_chalk2['default'].red(msg));
  process.exit(1);
}
